Slim/Web/SqueezeosTimezone.pm

```diff
@@ -0,0 +1,161 @@
+
+	my $res = eval { from_json($http->content) };
+	if ($@ || ref $res ne 'HASH') {
+		$log->error($@ || 'Invalid JSON response: ' . $http->content);
+		$res = {}; # guarantee that $res will be a hash ref
+	}
+
+	# '$tz' will hold the Olson formatted TimeZone to be returned.
+	# Setting '$tz' to '' signals a validation failure, and triggers a
+	# 404 error response.
+
+	my $tz = $res->{'timezone'};
+	if (!defined $tz || ref $tz) {
+		$log->error('Unexpected JSON response, expected a timezone string: ' . $http->content);
+		$tz = ''; # guarantee that $tz will be a string scalar
+	}
+
+	$tz = "$tz"; # ensure $tz is a string scalar
+	# Trim any leading/trailing white space, should it be there.
+	$tz =~ s/\A\s+|\s+\z//g;
+
+	$log->info("TimeZone query: Retrieved TimeZone \"$tz\"");
+	my $savedTz = $tz;
+
+	# Sanity check on TimeZone string.
+
+	# A TimeZone identifier is, essentially, a POSIX path with some
+	# additional (more and less voluntary) restrictions. These are
+	# indicated in the "theory" section of the tz distribution:
+	#   https://github.com/eggert/tz/blob/main/theory.html
+
+	# Path components should contain only A-Z, a-z, '.', '-', and '_'.
+	# And we need '/' (directory) to join the path components together.
+	# Note:
+	#  Some "legacy" and "etc" TimeZones may also contain 0-9 and '+', but
+	#  we do not expect or support such oddities.
+
+	$tz = '' if $tz =~ m{[^A-Za-z._\-/]} ; # reject if any characters outside that range
+
+	# Additional restrictions
+	$tz = '' if $tz =~ m{^/}; # leading '/' not allowed
+	$tz = '' if $tz =~ m{/$}; # trailing '/' not allowed
+	$tz = '' if $tz =~ m{//}; # no path component to be empty
+	$tz = '' if $tz =~ m{ ^- | /- }x; # no path component to start with a hyphen
+
+	# Path components that consist of singleton '.' or doubleton '..' are
+	# not allowed for obvious reasons.
+	# Other than that, the "theory" section of the tz distribution does
+	# allow "dots" elsewhere in TimeZone identifiers, but discourages
+	# them. That said, there are none defined at present, and almost
+	# certainly won't be. So just exclude any TimeZone containing a dot.
+	$tz = '' if $tz =~ m{\.}; # no path component to contain a '.'
+
+	# The 'Factory' TimeZone should never be encountered. SqueezeOS uses
+	# it to signal that the device's TimeZone has not been set.
+	# The 'Etc/Unknown' TimeZone signals an 'Unknown or Invalid' TimeZone,
+	# and is not defined in the tz database. SqueezeOS doesn't want it.
+	$tz = '' if $tz eq 'Factory';
+	$tz = '' if $tz eq 'Etc/Unknown';
+
+	# Note:
+	#  We do not guarantee to purge all invalid TimeZones with the above
+	#  sanity checks.
+
+	# All done, return result to SqueezeOS.
+	# But a 404 error if TimeZone failed validation.
+	unless ($tz eq '') {
+		$response->code(RC_OK);
+		$log->info("TimeZone query: Returning \"$tz\" to SqueezeOS device");
+	} else {
+		$response->code(RC_NOT_FOUND);
```

Comment:
I don't like the 404. Should we fall back to a default timezone instead? If the client can handle it I'd rather return a 204 (successfully executed, but no data), or even a 200 with an empty body.

Reply:
Now fixed in later commit.